lua/cmp_tabnine/source.lua

```diff
@@ -194,15 +207,40 @@ Source._on_stdout = function(_, data, _)
 			else
 				local results = response.results
 				old_prefix = response.old_prefix
+
 				if results ~= nil then
 					for _, result in ipairs(results) do
+						local newText = result.new_prefix .. result.new_suffix
+
+						local old_suffix = result.old_suffix
+						if string.sub(old_suffix, -1) == "\n" then
+							old_suffix = string.sub(old_suffix, 1, -2)
+						end
+
 						local item = {
-							label = result.new_prefix;
-							filterText = result.new_prefix;
-							insertText = result.new_prefix;
+							label = newText;
+							filterText = newText;
 							data = result;
-							sortText = result.new_prefix;
+							textEdit = {
+								range = {
+									start = { line = cursor.line, character = cursor.col - #old_prefix - 1 },
+									['end'] = {  line = cursor.line, character = cursor.col  + #old_suffix - 1 };
+								};
+								newText = result.new_prefix;
+							};
+							sortText = newText;
 						}
+						-- This is a hack fix for cmp not displaying items of TabNine::config_dir, version, etc. because their
+						-- completion items get scores of 0 in the matching algorithm
+						if #old_prefix == 0 then
+							item['filterText'] = string.sub(ctx.context.cursor_before_line, ctx.offset) .. newText
+						end
+
+						if #result.new_suffix > 0 then
+							item["insertTextFormat"] = cmp.lsp.InsertTextFormat.Snippet
+							item["textEdit"].newText = escape_tabstop_sign(result.new_prefix) .. '${0}' .. escape_tabstop_sign(result.new_suffix);
```

Comment:
shouldn't you also change the range in this case?

Reply:
and why treat this as a snippet?

Reply:
Replace range is only related to `old_prefix` and `old_suffix` and is already handled.

Reply:
snippet is used to keep the postion of cursor before `new_suffix` after confirmation. I borrowed this idea from https://github.com/codota/tabnine-vscode/blob/master/src/provideCompletionItems.ts.